operator/builtin/parser/csv/csv.go

```diff
@@ -87,42 +94,66 @@ func (c CSVParserConfig) Build(context operator.BuildContext) ([]operator.Operat
 // CSVParser is an operator that parses csv in an entry.
 type CSVParser struct {
 	helper.ParserOperator
-	header         []string
-	fieldDelimiter rune
-	numFields      int
-	lazyQuotes     bool
+	fieldDelimiter  rune
+	header          []string
+	headerAttribute string
+	lazyQuotes      bool
+	parse           parseFunc
 }
 
-// Process will parse an entry for csv.
-func (r *CSVParser) Process(ctx context.Context, entry *entry.Entry) error {
-	return r.ParserOperator.ProcessWith(ctx, entry, r.parse)
-}
-
-// parse will parse a value using the supplied csv header.
-func (r *CSVParser) parse(value interface{}) (interface{}, error) {
-	var csvLine string
-	switch val := value.(type) {
-	case string:
-		csvLine = val
-	default:
-		return nil, fmt.Errorf("type '%T' cannot be parsed as csv", value)
-	}
-
-	reader := csvparser.NewReader(strings.NewReader(csvLine))
-	reader.Comma = r.fieldDelimiter
-	reader.FieldsPerRecord = r.numFields
-	reader.LazyQuotes = r.lazyQuotes
-	parsedValues := make(map[string]interface{})
-
-	record, err := reader.Read()
+type parseFunc func(interface{}) (interface{}, error)
 
-	if err != nil {
-		return nil, err
+// Process will parse an entry for csv.
+func (r *CSVParser) Process(ctx context.Context, e *entry.Entry) error {
+	if r.headerAttribute != "" {
+		h, ok := e.Attributes[r.headerAttribute]
+		if !ok {
+			err := fmt.Errorf("failed to read dynamic header attribute %s", r.headerAttribute)
+			r.Error(err)
+			return err
+		}
+		headers := strings.Split(h, string([]rune{r.fieldDelimiter}))
+		r.parse = generateParseFunc(headers, r.fieldDelimiter, r.lazyQuotes)
 	}
+	return r.ParserOperator.ProcessWith(ctx, e, r.parse)
+}
 
-	for i, key := range r.header {
-		parsedValues[key] = record[i]
+// generateParseFunc returns a parse function for a given header, allowing
+// each entry to have a potentially unique set of fields when using dynamic
+// field names retrieved from an entry's attribute
+func generateParseFunc(headers []string, fieldDelimiter rune, lazyQuotes bool) parseFunc {
+	return func(value interface{}) (interface{}, error) {
+		var csvLine string
+		switch t := value.(type) {
+		case string:
+			csvLine += t
+		case []byte:
+			csvLine += string(t)
+		default:
+			return nil, fmt.Errorf("type '%T' cannot be parsed as csv", value)
+		}
+
+		reader := csvparser.NewReader(strings.NewReader(csvLine))
+		reader.Comma = fieldDelimiter
+		reader.FieldsPerRecord = len(headers)
+		reader.LazyQuotes = lazyQuotes
+		parsedValues := make(map[string]interface{})
+
+		for {
+			body, err := reader.Read()
+			if err == io.EOF {
+				break
+			}
+
+			if err != nil {
+				return nil, err
+			}
+
+			for i, key := range headers {
+				parsedValues[key] = body[i]
+			}
+		}
+
+		return parsedValues, nil
```

Comment:
It might be a little cleaner to seperate all the dynamic header related logic into a receiver on the operator. You don't have to pass a bunch of operator fields around, and can just generate, use, and forget the func all in one place.
```go
func (r *CSVParser) Process(ctx context.Context, e *entry.Entry) error {
	if r.parse != nil {
		return r.ParserOperator.ProcessWith(ctx, e, r.parse)
	}
	return r.dynamicParse(ctx, e)
}
```

```suggestion
func (r *CSVParser) dynamicParse(ctx context.Context, e *entry.Entry)  error {
	h, ok := e.Attributes[r.headerAttribute]
	if !ok {
		err := fmt.Errorf("failed to read dynamic header attribute %s", r.headerAttribute)
		r.Error(err)
		return err
	}
	headers := strings.Split(h, string([]rune{r.fieldDelimiter}))
	
	return r.ParserOperator.ProcessWith(ctx, e, 	
		func(value interface{}) (interface{}, error) {
			...
			reader := csvparser.NewReader(strings.NewReader(csvLine))
			reader.Comma = r.fieldDelimiter
			reader.FieldsPerRecord = len(headers)
			reader.LazyQuotes = r.lazyQuotes
			parsedValues := make(map[string]interface{})

			for {
				body, err := reader.Read()
				if err == io.EOF {
					break
				}

				if err != nil {
					return nil, err
				}

				for i, key := range headers {
					parsedValues[key] = body[i]
				}
			}

			return parsedValues, nil
		}
	)
```

Reply:
I tried this but I'm not sure I like it. You're right we do use some fields from the struct itself but also it means we regenerate the parse function every time. For configurations that are not using dynamic parsing (which is likely most of the time) we're just wasting processing time recalculating the same function over and over for every entry.

It is a bit ugly to have generate parse take in fields we have on the struct but also this way we're optimizing for the more common case instead of the edge case.

Reply:
> For configurations that are not using dynamic parsing (which is likely most of the time) we're just wasting processing time recalculating the same function over and over for every entry.

Not sure why we would do that. I don't think I suggested to do it that way. This func would only be called in the same circumstances where we currently call `generateParseFunc`.

Reply:
To clarify this earlier comment:
```go
func (r *CSVParser) Process(ctx context.Context, e *entry.Entry) error {

	// r.parse set in Build function if using constant headers
	if r.parse != nil {
		return r.ParserOperator.ProcessWith(ctx, e, r.parse)
	}

	// only runs in dynamic headers case
	return r.dynamicParse(ctx, e)
}
```

Reply:
Sorry, I think I was operating under the assumption that the above issue didn't cause a race condition. This makes more sense now with that and I agree with your change I'll update it.

Reply:
The only issue with this is the initialization of the `parse` field on the struct is basically the `generateParseFunc` function we have now. By condensing all the parse logic into a receiver on the struct It makes it so you need to have an anonymous function inside `dynamicParse` in order to generate your parsing function.

I would propose keeping `generateParseFunc` and mixing it with the above code where if `r.parse` is nil return `r.ParserOperator.ProcessWith(ctx, e, generateParseFunc)`. Cleaned up but that's the gist.

Reply:
I pushed up a small change that avoids saving the dynamic function. Let me know if you still feel strongly that it should go into a receiver.